### ISSUE TITLE
mSD复现结果

### ISSUE THREAD

**Issue by dddcaiji** [Zplusdragon/UFineBench#9]
![Image](https://github.com/user-attachments/assets/c8768e32-77e8-4d51-9eee-31147abb8962)

IRRA的mSD指标复现结果差距好远

**Comment by dddcaiji** [comment]
def mSD_rank(similarity, q_pids, g_pids, cost_matrix=None, max_rank=10):
    similarity = similarity / 2 + 0.5
    similarity = similarity.cpu()
    indices = torch.argsort(similarity, dim=1, descending=True)
    indices = indices.cpu()
    pred_labels = g_pids[indices.cpu()]  # q * k
    matches = pred_labels.eq(q_pids.view(-1, 1))  # q × k

    all_cmc = matches[:, :max_rank].cumsum(1)
    all_cmc[all_cmc > 1] = 1
    all_cmc = all_cmc.float().mean(0) * 100

    # 正负样本相似度
    postive_idx = (indices.cpu() + 1) * matches
    postive_idx[postive_idx >= 1] = 1
    postive_similarity = similarity.cpu() * postive_idx

    num_rel = matches.sum(1)
    postive_similarity_sum = postive_similarity.sum(1)
    negative_similarity_sum = similarity.sum(1) - postive_similarity_sum
    postive_similarity_average = postive_similarity_sum / num_rel
    negative_similarity_average = negative_similarity_sum / (similarity.shape[1] - num_rel)
    pn_ratio = (postive_similarity_average / negative_similarity_average).numpy()
    pn_ratio = torch.tensor([1 - math.exp(-x) for x in pn_ratio])
    similarity_cmc = torch.cumsum(similarity.cpu(), dim=1)
    positive_similarity_cmc = torch.cumsum(postive_similarity, dim=1)

    sd_cmc = positive_similarity_cmc / similarity_cmc
    sd_cmc = sd_cmc * matches

    SD = (sd_cmc.sum(1) / num_rel) * pn_ratio
    print((sd_cmc.sum(1) / num_rel))
    print(pn_ratio)
    print("SD:", SD.device)
    mSD = SD.mean() * 100

    # mAP 保持不变（用于对比）
    tmp_cmc = matches.cumsum(1)
    tmp_cmc = [tmp_cmc[:, i] / (i + 1.0) for i in range(tmp_cmc.shape[1])]
    tmp_cmc = torch.stack(tmp_cmc, 1) * matches
    AP = tmp_cmc.sum(1) / num_rel
    mAP = AP.mean() * 100

    return all_cmc, mSD, mAP
